gh-99631: Add custom loads and dumps support for the shelve module (#118065)
python/cpython: Doc/library/shelve.rst

   determine which accessed entries are mutable, nor which ones were actually
   mutated).

   By default, :mod:`shelve` uses :func:`pickle.dumps` and :func:`pickle.loads`

>    for serializing and deserializing. However *serializer* can be the function

Reviewer: ```suggestion
   for serializing and deserializing. However *deserializer* can be the function
```

>    that takes the :term:`bytes-like object` and returns the object. *deserializer*

Reviewer: ```suggestion
   that takes the :term:`bytes-like object` and returns the object. *serializer*
```

   can be the function that takes the object and returns :class:`bytes`. For example,
   :keyword:`lambda`, which the :mod:`pickle` does not support, can be used in
   :mod:`shelve` using the serializer and deserializer functions, which do support

>    the :keyword:`lambda`.

Reviewer 2: Text needs a little grammar polish

>    the :keyword:`lambda`.

Reviewer: The versionchanged statement below needs to be adapted

   determine which accessed entries are mutable, nor which ones were actually
   mutated).

   By default, :mod:`shelve` uses :func:`pickle.dumps` and :func:`pickle.loads`
   for serializing and deserializing. This can be changed by supplying
   *serializer* and *deserializer*, respectively. The *serializer* argument

>    should be a function that takes an object and returns its representation

Reviewer 3: Mention that the function must also take a protocol argument that may be ignored, namely, it will be called as

```py
data = serializer(obj, protocol)
```

Author: > 
The *serializer* argument should be a function that takes an object and  the *protocol* argument passed to the open function and returns its representation as a :term:`bytes-like object`; *protocol* argument that may be ignored by the function.

But it might be a bit confusing, because the `serializer` function [uses](https://github.com/python/cpython/pull/118065/files#diff-3a8b6356369487dc852c6841e6710e162a5e740b8fefdbcfc3df4fe834e756cbR137) the `protocol` argument passed to the open function.

What do you think about this? @picnixz

Reviewer 3: I think your suggestion is fine because the serializer may depend on the protocol or not and yes, it should be like this IMO. I mean, maybe your serializer does something different depending on the protocol itself, like patching some parts. Therefore, I don't think it would be "confusing", but if you want to make it less confusing, just mention how the serializer is being called (a good example is worth a thousand words).

   as a :term:`bytes-like object`; *deserializer* should be a function that

>    takes :class:`bytes` and returns the corresponding object.

Reviewer 3: `pickle.loads` accepts bytes-like objects as well. If you accept arbitrary bytes objects as the `serializer` output, you should also test the various objects implementing the buffer protocol.

>    If one of these is given, the other must be given as well.

Reviewer 3: Maybe mention that an exception is being raised otherwise?

>    By default, :mod:`shelve` uses :func:`pickle.dumps` and :func:`pickle.loads`
>    for serializing and deserializing. However *deserializer* can be the function
>    that takes the :term:`bytes-like object` and the *protocol* parameter and
>    returns the object. *serializer* can be the function that takes the object
>    and returns :class:`bytes`. For example, :keyword:`lambda`, which
>    the :mod:`pickle` does notsupport, can be used in :mod:`shelve` using
>    theserializer and deserializer functions, which do support the
>    :keyword:`lambda`.

Reviewer 4: - The arguments are in the wrong order.
- They need to be defined together
- IMO, the cloudpickle/dill use case can be omitted. There are several other ones, like *limiting* the acceptable types (e.g. using `json` for better security).

How about:
```suggestion
   By default, :mod:`shelve` uses :func:`pickle.dumps` and :func:`pickle.loads`
   for serializing and deserializing. This can be changed by supplying
   *serializer* and *deserializer*, respectively. The *serializer* argument
   should be a function that takes an object and returns its representation
   as a :term:`bytes-like object`; *deserializer* should be a function that
   takes :class:`bytes` and returns the corresponding object.
   If one of these is given, the other must be given as well.
```

   determine which accessed entries are mutable, nor which ones were actually
   mutated).

   By default, :mod:`shelve` uses :func:`pickle.dumps` and :func:`pickle.loads`
   for serializing and deserializing. This can be changed by supplying
   *serializer* and *deserializer*, respectively. The *serializer* argument
   should be a function that takes an object and the *protocol* argument passed
   to the open function and returns its representation as a
   :term:`bytes-like object`; *protocol* argument that may be ignored by the
   function. *deserializer* should be a function that takes :class:`bytes` and
   returns the corresponding object. If one of these is given, the other must

>    be given as well. Otherwise :mod:`shelve` will raise a :exc:`ShelveError`.

Reviewer 3: ```suggestion
   By default, :mod:`shelve` uses :func:`pickle.dumps` and :func:`pickle.loads`
   for serializing and deserializing. This can be changed by supplying
   *serializer* and *deserializer*, respectively.

   The *serializer* argument must be a callable which takes an object ``obj``
   and the *protocol* as inputs and returns the representation ``obj`` as a
   :term:`bytes-like object`; the *protocol* value may be ignored by the
   serializer.

   The *deserializer* argument must be callable which takes a serialized object
   given as a :class:`bytes` object and returns the corresponding object.

   A :exc:`ShelveError` is raised if *serializer* is given but *deserializer*
   is not, or vice-versa.
```

   .. versionchanged:: 3.10
      :const:`pickle.DEFAULT_PROTOCOL` is now used as the default pickle
      protocol.

   .. versionchanged:: 3.11
      Accepts :term:`path-like object` for filename.

>    .. versionchanged:: 3.15

Reviewer 3: ```suggestion
   .. versionchanged:: next
```

  which can cause hard crashes when trying to read from the database.

.. class:: Shelf(dict, protocol=None, writeback=False, keyencoding='utf-8', \

>                  serializer=None, deserializer=None)

Reviewer 3: I think the *serializer* and *deserializer* parameters must be keyword-only, so to reflect the usage of `shelve.open`.

  which can cause hard crashes when trying to read from the database.

.. class:: Shelf(dict, protocol=None, writeback=False, \

>                  keyencoding='utf-8', * serializer=None, deserializer=None)

Reviewer 3: ```suggestion
                 keyencoding='utf-8', *, serializer=None, deserializer=None)
```

   The *keyencoding* parameter is the encoding used to encode keys before they
   are used with the underlying dict.

>    The *deserializer* parameter can be the function that takes the
>    :term:`bytes-like object` and the *protocol* parameter and returns the
>    object. *serializer* parameter can be the function that takes the object
>    and returns :class:`bytes`.
> 

Reviewer 4: These only need one description:
```suggestion
   The *deserializer* and *serializer* are as in :func:`~shelve.open`.

```
(If you accept this, also replace `Shelve` by `shelve.open` in `BsdDbShelf` & `DbfilenameShelf` below.)

   The *keyencoding* parameter is the encoding used to encode keys before they
   are used with the underlying dict.

>    The *deserializer* and *serializer* are as in :func:`~shelve.open`.

Reviewer 3: ```suggestion
   The *serializer* and *deserializer* parameters have the same interpretation
   as in :func:`~shelve.open`.
```

To have consistent wording.

      :const:`pickle.DEFAULT_PROTOCOL` is now used as the default pickle
      protocol.

   .. versionchanged:: 3.14

>       Accepts custom *serializer* and *deserializer* functions in place of
>       :func:`pickle.dumps` and :func:`pickle.loads`.

Reviewer 3: ```suggestion
      Added the *serializer* and *deserializer* parameters.
```

I think you do not need to duplicate what you already said in `shelve.open`.

.. class:: BsdDbShelf(dict, protocol=None, writeback=False, \

>                       keyencoding='utf-8',  serializer=None, deserializer=None)

Reviewer 3: Same comment on being keyword-only.

      :const:`pickle.DEFAULT_PROTOCOL` is now used as the default pickle
      protocol.

   .. versionchanged:: 3.14

>       Added the *serializer* and *deserializer* parameters.

Reviewer 3: Could you keep two blank lines between above the `.. class::`? I think there is only one here.

      :const:`pickle.DEFAULT_PROTOCOL` is now used as the default pickle
      protocol.

>    .. versionchanged:: 3.15

Reviewer 3: ```suggestion
   .. versionchanged:: next
```

   modules.  The *dict* object passed to the constructor must support those
   methods.  This is generally accomplished by calling one of
   :func:`!bsddb.hashopen`, :func:`!bsddb.btopen` or :func:`!bsddb.rnopen`.  The
   optional *protocol*, *writeback*, *keyencoding*, *serializer* and *deserializer*
   parameters have the same interpretation as for the :func:`~shelve.open`.

.. class:: DbfilenameShelf(filename, flag='c', protocol=None, writeback=False, \

>                            serializer=None, deserializer=None)

Reviewer 3: Ditto

   A subclass of :class:`Shelf` which accepts a *filename* instead of a dict-like
   object.  The underlying file will be opened using :func:`dbm.open`.  By
   default, the file will be created and opened for both read and write.  The
   optional *flag* parameter has the same interpretation as for the
   :func:`.open` function.  The optional *protocol*, *writeback*, *serializer*
   and *deserializer* parameters have the same interpretation as for the

>    :func:`~shelve.open`

Reviewer 3: ```suggestion
   :func:`~shelve.open`.
```

   modules.  The *dict* object passed to the constructor must support those
   methods.  This is generally accomplished by calling one of

>    :func:`!bsddb.hashopen`, :func:`!bsddb.btopen` or :func:`!bsddb.rnopen`.  The

Author: By the way, we also need to update this sentence (from `bsddb` to `berkeleydb`). `bsddb` is deprecated according to https://www.jcea.es/programacion/pybsddb.htm.

Reviewer 2: It looks like the change would be bigger: it’s a new module (although the docs are not very clear) with maybe a new API.

Updating or deprecating this should be discussed in its own ticket 🙂

Author: Yes, I agree on it. It would be better to open a new ticket to discuss this issue.

Reviewer 5: ... or consider instead opening topic on Discourse; this may need a larger audience than what you'd get on the bug tracker.

   modules.  The *dict* object passed to the constructor must support those
   methods.  This is generally accomplished by calling one of
   :func:`!bsddb.hashopen`, :func:`!bsddb.btopen` or :func:`!bsddb.rnopen`.  The
   optional *protocol*, *writeback*, *keyencoding*, *serializer* and *deserializer*

>    parameters have the same interpretation as for the :func:`~shelve.open`.

Reviewer 3: ```suggestion
   parameters have the same interpretation as in :func:`~shelve.open`.
```

   .. versionchanged:: 3.14
      Added the *serializer* and *deserializer* parameters.

.. class:: DbfilenameShelf(filename, flag='c', protocol=None, \
                           writeback=False, *, serializer=None, \
                           deserializer=None)

   A subclass of :class:`Shelf` which accepts a *filename* instead of a dict-like
   object.  The underlying file will be opened using :func:`dbm.open`.  By
   default, the file will be created and opened for both read and write.  The
   optional *flag* parameter has the same interpretation as for the
   :func:`.open` function.  The optional *protocol*, *writeback*, *serializer*

>    and *deserializer* parameters have the same interpretation as for the

Reviewer 3: ```suggestion
   and *deserializer* parameters have the same interpretation as in
```

   modules.  The *dict* object passed to the constructor must support those
   methods.  This is generally accomplished by calling one of
   :func:`!bsddb.hashopen`, :func:`!bsddb.btopen` or :func:`!bsddb.rnopen`.  The
   optional *protocol*, *writeback*, *keyencoding*, *serializer* and *deserializer*
   parameters have the same interpretation as in :func:`~shelve.open`.

   .. versionchanged:: 3.14

>       Added the *serializer* and *deserializer* parameters.

Reviewer 3: Ditto

.. class:: DbfilenameShelf(filename, flag='c', protocol=None, \
                           writeback=False, *, serializer=None, \
                           deserializer=None)

   A subclass of :class:`Shelf` which accepts a *filename* instead of a dict-like
   object.  The underlying file will be opened using :func:`dbm.open`.  By
   default, the file will be created and opened for both read and write.  The
   optional *flag* parameter has the same interpretation as for the
   :func:`.open` function.  The optional *protocol*, *writeback*, *serializer*
   and *deserializer* parameters have the same interpretation as in the
   :func:`~shelve.open`.

   .. versionchanged:: 3.14

>       Added the *serializer* and *deserializer* parameters.

Reviewer 3: Ditto

   modules.  The *dict* object passed to the constructor must support those
   methods.  This is generally accomplished by calling one of
   :func:`!bsddb.hashopen`, :func:`!bsddb.btopen` or :func:`!bsddb.rnopen`.  The
   optional *protocol*, *writeback*, *keyencoding*, *serializer* and *deserializer*
   parameters have the same interpretation as in :func:`~shelve.open`.

>    .. versionchanged:: 3.15

Reviewer 3: ```suggestion
   .. versionchanged:: next
```

      Added the *serializer* and *deserializer* parameters.

.. class:: DbfilenameShelf(filename, flag='c', protocol=None, \
                           writeback=False, *, serializer=None, \
                           deserializer=None)

   A subclass of :class:`Shelf` which accepts a *filename* instead of a dict-like
   object.  The underlying file will be opened using :func:`dbm.open`.  By
   default, the file will be created and opened for both read and write.  The
   optional *flag* parameter has the same interpretation as for the
   :func:`.open` function.  The optional *protocol*, *writeback*, *serializer*

>    and *deserializer* parameters have the same interpretation as in the

Reviewer 3: ```suggestion
   and *deserializer* parameters have the same interpretation as in
```

   :func:`~shelve.open`.

>    .. versionchanged:: 3.15

Reviewer 3: ```suggestion
   .. versionchanged:: next
```

   d.close()                  # close it

Exceptions
----------

.. exception:: ShelveError

   Exception raised when one of the arguments *deserializer* and *serializer*
   is missing in the :func:`~shelve.open`, :class:`Shelf`, :class:`BsdDbShelf`
   and :class:`DbfilenameShelf`

   The *deserializer* and *serializer* arguments must be given together.

   .. versionadded:: 3.14

> 

Reviewer 3: Ditto

   d.close()                  # close it

Exceptions
----------

.. exception:: ShelveError

   Exception raised when one of the arguments *deserializer* and *serializer*
   is missing in the :func:`~shelve.open`, :class:`Shelf`, :class:`BsdDbShelf`

>    and :class:`DbfilenameShelf`

Reviewer 3: ```suggestion
   and :class:`DbfilenameShelf`.
```

   The *deserializer* and *serializer* arguments must be given together.

>    .. versionadded:: 3.15

Reviewer 3: ```suggestion
   .. versionadded:: next
```
